concordium-node/src/p2p/peers.rs

```diff
@@ -169,9 +175,143 @@ fn calculate_average_throughput(
+                let ip = source.read_u32::<BigEndian>()?;
+                let mut buf = [0; 4];
+                BigEndian::write_u32(&mut buf, ip);
+                Self(SocketAddr::new(std::net::IpAddr::V4(buf.into()), port))
+            }
+            1 => {
+                let port = source.read_u16::<BigEndian>()?;
+                let ip = source.read_u128::<BigEndian>()?;
+                let mut buf = [0; 16];
+                BigEndian::write_u128(&mut buf, ip);
+                Self(SocketAddr::new(std::net::IpAddr::V6(buf.into()), port))
+            }
+            _ => anyhow::bail!("Unsupported type of `StoredPeer`"),
+        };
+        Ok(peer)
+    }
+}
+
+/// Persist the [`SocketAddr`] of a peer.
+pub fn persist_peer(node: &Arc<P2PNode>, peer_addr: SocketAddr) -> anyhow::Result<()> {
+    if let Ok(kv) = node.kvs.read() {
+        let peers_store = kv.open_single(PEERS_STORE_NAME, StoreOptions::create())?;
+        let mut buf = Vec::new();
+        let stored_peer: StoredPeer = peer_addr.into();
+        stored_peer.serial(&mut buf);
+        let mut writer = kv.write()?;
+        peers_store.put(&mut writer, buf, &Value::U64(0))?;
+        writer.commit()?;
+        Ok(())
+    } else {
+        anyhow::bail!("Could not acqure lock over lmdb");
+    }
+}
+
+/// Remove a peer from the persisted peer database.
+pub fn remove_persisted_peer(node: &Arc<P2PNode>, peer_addr: SocketAddr) -> anyhow::Result<()> {
+    if let Ok(kv) = node.kvs.read() {
+        let peers_store = kv.open_single(PEERS_STORE_NAME, StoreOptions::create())?;
+        let mut key = Vec::new();
+        let stored_peer: StoredPeer = peer_addr.into();
+        stored_peer.serial(&mut key);
+        let mut writer = kv.write()?;
+        peers_store.delete(&mut writer, key)?;
+        writer.commit()?;
+        Ok(())
+    } else {
+        anyhow::bail!("Could not acqure lock over lmdb");
+    }
+}
+
+/// Try connect to previosly connected peers if
+/// anyone is stored.
+/// Note that as opposed to [`connect_to_config_nodes`] this function respects
+/// the maximum peers configured for the node.
+pub fn connect_to_stored_nodes(node: &Arc<P2PNode>) -> anyhow::Result<()> {
+    if let Ok(kvs_env) = node.kvs.read() {
+        let peers_store = kvs_env.open_single(PEERS_STORE_NAME, StoreOptions::create())?;
+        let peers_reader = kvs_env.read()?;
+        let peers_iter = peers_store.iter_start(&peers_reader)?;
+        for entry in peers_iter {
+            let (mut peer_bytes, _expiry) = entry?;
+            if let Err(e) =
+                connect(node, PeerType::Node, StoredPeer::deserial(&mut peer_bytes)?.0, None, true)
+            {
+                warn!("could not connect to previosly connected peer {}", e);
+            }
+        }
+        Ok(())
+    } else {
+        anyhow::bail!("could not read previosly connected peers: cannot obtain lock for the kvs");
+    }
+}
+
+impl P2PNode {
```

Comment:
Is there a reason for not having the above functions as a method on P2PNode? It is a bit inconsistent since `clear_persisted_peers` is defined as a method.

Reply:
Moved `persist_peer` and  `remove_persisted_peer` to `Impl P2PNode`.